src/amd_sgpio.h

```diff
@@ -18,8 +18,33 @@
  *
  */
 
+#ifndef _AMD_SGPIO_INCLUDED_
+#define _AMD_SGPIO_INCLUDED_
+
 #include "block.h"
+#include "libled_internal.h"
+
+#include <stdint.h>
+
+typedef uint8_t drive_led_t;
+
+struct drive_leds {
+	drive_led_t	error;
+	drive_led_t	locate;
+	drive_led_t	activity;
+} __attribute__ ((__packed__));
 
-int _amd_sgpio_em_enabled(const char *path);
+struct cache_entry {
+	struct drive_leds leds[4];
+	uint8_t blink_gen_a;
+	uint8_t blink_gen_b;
+	uint16_t reserved;
+} __attribute__ ((__packed__));
+
+int _amd_sgpio_em_enabled(const char *path, struct led_ctx *ctx);
 int _amd_sgpio_write(struct block_device *device, enum ibpi_pattern ibpi);
-char *_amd_sgpio_get_path(const char *cntrl_path);
+char *_amd_sgpio_get_path(const char *cntrl_path, struct led_ctx *ctx);
+
+void amd_sgpio_cache_free(struct led_ctx *ctx);
+
+#endif
```

Comment:
> Perhaps, we should consider new method in libled.h slot_get_controller to avoid referencing slot content directly? I've put comment there too.

We have a function to retrieve this information `led_slot_cntrl`.  I unfortunately used the internal representation as we have access to it.  Will refactor so this doesn't work anymore.

Reply:
done